arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/NonEmptyCollection.kt

```diff
@@ -0,0 +1,21 @@
+package arrow.core
+
+/**
+ * Common interface for collections that always have
+ * at least one element (available from [head]).
+ */
+public interface NonEmptyCollection<out E> : Collection<E> {
+  override fun isEmpty(): Boolean = false
+  public val head: E
```

Comment:
I saw in the Slack thread you mentioned `first()`. Wondering if it's worth deprecating `head` to `first()` or if that would cause confusion with the Kotlin Std method `first` that throws `OutOfBoundsException` 🤔

Reply:
The reason I didn't do this is because `first` is defined as an extension method, and I was a bit unsure of how the shadowing of those methods would work. And in fact, using the one in `kotlin.collections` is fine, we just know we don't get an exception...

Something more interesting would be overriding `firstOrNull()` with a non-nullable type, but I'm not sure how many people are in the situation of using `firstOrNull()` on `NonEmptyCollection` in the first place.

Reply:
Ah right, those are extension functions defined in the Kotlin Std... If it's a method, then it should always take precedence over any extension functions but perhaps it's better not to conflict at all. `first` will in this case be "safe" anyway, so let's keep it as is.

We can override `firstOrNull` as an instance method, and it will take precedence over the `Iterable#firstNotNull` method from Kotlin Std.

Reply:
I've just pushed another commit where a bunch of functions are overridden to track for non-emptiness:
- `firstOrNull`, `lastOrNull`,
- `map`,
- `distinct` and `distinctBy`,
- `flatMap` and `zip` when the argument is also non-empty.